doc/stm/index.mld

```diff
@@ -0,0 +1,581 @@
+
+  let arb_cmd _state =
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Cardinal;
+         Gen.map (fun i -> Mem i) Gen.int;
+         Gen.map (fun i -> Add i) Gen.int;
+        ])
+end
+]}
+
+Let's go over the content of that specification.
+
+The first thing we do is to instantiate the functor. We can't test the library
+if we don't. In the same way, if you want to test a library with an abstract
+type parameter (like ['a t]), you will have to instantiate the ['a].
+
+{3 The library description}
+
+[sut] is the main type of the system under test. Here the mutable set.
+[qcheck-stm] also needs to know how to create an initial value and how to
+clean up after the tests (which is necessary when the [sut] uses resources that
+must be released, such as opened files, network connections, etc.).
+
+The type [cmd] describes the functions of the library we want to test. The
+arguments of the constructors correspond to the arguments of the corresponding
+functions except for the [sut]. Note that we use a ppx_deriver to generate a
+[show] function. We will need it in [arb_cmd]. But you can also implement it
+yourself if you prefer.
+
+Finally, the [run] function calls the library's function with the given arguments
+and [sut] and wraps the result with information about its type in a
+{{!STM_base.res}res} type.
+
+{3 The model definition}
+
+[state] is the type of the model. The model should be pure and simple enough to
+be obviously correct. Here we use a simple {{!Stdlib.List}list} from the
+standard library.
+
+[qcheck-stm] also needs to know what is the initial value. This is provided by
+[init_state] and should be the model of the result of [init_sut ()].
+
+The function [next_state] updates the [state] according to a
+[cmd]. Here, only the [Add] [cmd] changes the [state]. The result of running a
+[cmd] will be checked in [postcond].
+
+The [precond] function checks whether the [cmd] can be run. Here, it is always
+the case. But this is useful when some library functions have preconditions.
+
+The [postcond] function is the last concerning the model against which the
+library will be tested. It takes a [cmd], the [state] {i before} running the
+[cmd] and the [res] returned by the [run] function. In this function, we
+express the relation between the result of the library's functions and the
+functional model.  The fact that it takes the [state] before the call of the
+function will come in handy when we will add a {{!section-remove}[remove]}
+function to our set library.
+
+{3 The [cmd]s generation}
+
+The last thing to do is to tell [qcheck-stm] how to generate the [cmd]s. This
+is what is done by the [arb_cmd] function, using [QCheck]'s combinators. We
+will make two remarks on this function.  The first one is that [QCheck.make]
+takes an optional printer. It is important to provide it so that the test's
+output is readable.  Here we use the function [show_cmd] that has been built by
```

Comment:
'readable' may sound as if it is otherwise unreadable in the sense of being "garbled".
If that is the case, we should open an issue and fix it.
It not, I think that writing *"so that the test's output is printed"* is clearer.